wire-ios-data-model/Source/MLS/MLSService.swift

```diff
@@ -828,32 +819,33 @@ public final class MLSService: MLSServiceInterface {
         guard let context = context else {
             return
         }
-        let pendingJoins = await context.perform { self.generatePendingJoins(in: context) }
-        for pendingJoin in pendingJoins {
-            try await joinByExternalCommit(groupID: pendingJoin.groupID)
-        }
 
-        groupsPendingJoin.removeAll()
-    }
-
-    private func generatePendingJoins(in context: NSManagedObjectContext) -> [PendingJoin] {
-        logger.info("generating list of groups pending join")
+        let pendingGroups = await context.perform { self.fetchPendingGroups(in: context) }
 
-        return groupsPendingJoin.compactMap { groupID in
+        logger.info("joining \(pendingGroups.count) group(s)")
 
-            guard let conversation = ZMConversation.fetch(with: groupID, in: context) else {
-                logger.warn("conversation not found for group (\(groupID.safeForLoggingDescription))")
-                return nil
+        await withTaskGroup(of: Void.self) { group in
+            for pendingGroup in pendingGroups {
+                group.addTask {
+                    do {
+                        try await self.joinByExternalCommit(groupID: pendingGroup)
+                    } catch {
+                        WireLogger.mls.error("Failed to join pending group (\(pendingGroup): \(error)")
+                    }
+                }
             }
+        }
+    }
 
-            guard let status = conversation.mlsStatus, status == .pendingJoin else {
-                logger.warn("group (\(groupID.safeForLoggingDescription)) status is not pending join")
-                return nil
-            }
+    private func fetchPendingGroups(in context: NSManagedObjectContext) -> [MLSGroupID] {
+        logger.info("fetching list of groups pending join")
 
-            return (groupID, conversation.epoch)
+        let pendingConversations = ZMConversation.fetchConversationsWithMLSGroupStatus(
+            mlsGroupStatus: .pendingJoin,
+            in: context
+        )
 
-        }
+        return pendingConversations.compactMap(\.mlsGroupID)
```

Comment:
One could write this also in a functional like this, but I think it's personal preference 😉
```swift
return ZMConversation.fetchConversationsWithMLSGroupStatus(
    mlsGroupStatus: .pendingJoin,
    in: context
)
.compactMap(\.mlsGroupID)
```

Reply:
Done